README.rst
==========================
``gs.content.js.required``
==========================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Make required widgets actually required
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2010-10-10
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.Net`_.


Introduction
============

This product provides a JavaScript resource_ to prevent a form
from being submitted if the required widgets are not filled
out. The JavaScript is activated if the requisite markup_ is
present.

Resource
========

This product provides a JavaScript module as a Zope_ `browser
resource`_. Any Zope or Plone_ project should be able to use this
product as-is by placing the following line in a page template:

.. code-block:: xml

  <script type="text/javascript" 
          src="/++resource++gs-content-js-required-20130517.js" 
          defer="true"> </script>     

Users of other systems are invited to copy the file
``gs/content/js/required/browser/javascript/required.js`` out of this
product. 

A minified version of the script is available, too:

.. code-block:: xml

  <script type="text/javascript" 
          src="/++resource++gs-content-js-required-min-20130517.js" 
          defer="true"> </script>     

Markup
======

The markup that triggers the required-widgets JavaScript is in three parts:
the form, the widgets, and the button.

#.  The form should have the class ``gs-content-js-required``:

    .. code-block:: xml

      <form id="login-form" class="gs-content-js-required"
            tal:attributes="action request/URL">

#.  Widgets that are required should have the class ``required``
    surrounding the ``div``:


    .. code-block:: xml

      <div id="login-id-widget" class="form-widget required">
        <label class="text" for="login">Email Address</label>
        <input id="login" name="login" class="textType" type="text" 
               value="" title="Enter your email address here" />
      </div><!--login-id-widget-->

    This is automatically done by widgets generated by ``zope.formlib``.

#.  If there is only one button on the form, or all buttons on the form
    need to be locked, then nothing else has to be done. The button will be
    disabled until all the required widgets are filled out. If there are
    multiple buttons then the buttons to lock can be specified in the
    ``data-required-buttons`` attribute of the form:


    .. code-block:: xml

      <form id="add-form" class="gs-content-js-required"
            tal:attributes="action request/URL"
            data-required-buttons="#form\.add">

Resources
=========

- Code repository: https://github.com/groupserver/gs.content.js.required/
- Questions and comments to http://groupserver.org/groups/development/
- Report bugs at https://redmine.iopen.net/projects/groupserver/

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17
.. _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/
.. _Zope: http://zope.org/
.. _formlib: http://docs.zope.org/zope.formlib/
.. _browser resource: http://docs.zope.org/zope.browserresource/
.. _Plone: http://plone.org
.. _jQuery: http://jquery.com/

.. LocalWords:  jQuery UI Plone minified


docs/HISTORY.rst
Changelog
=========

2.1.3 (2015-11-06)
------------------

* Adding an ``interval`` to poll for a changed form-value
  `because Google Chrome`_ (see also `Detecting browser
  autofill`_ on Stack Overflow)

.. _because Google Chrome:
   https://code.google.com/p/chromium/issues/detail?id=352527

.. _Detecting browser autofill:
   http://stackoverflow.com/questions/11708092/detecting-browser-autofill
   
2.1.2 (2014-10-10)
------------------

* Switching to GitHub as the canonical repository
* Naming the reStructuredText files as such

2.1.1 (2014-07-14)
------------------

* Fixing an error in a variable name

2.1.0 (2014-03-13)
------------------

* Fixing an error with Microsoft Internet Explorer 8
* Switching to ``"use strict";``

2.0.0 (2013-05-22)
------------------

* Initial version, based on the old code in ``Products.GSContent``


version.py
version='2'
release=False

#-------------------------------------------------------------------------------------#

import commands
import datetime
import os
import glob

class CommandError(Exception):
    pass

def execute_command(commandstring):
    status, output = commands.getstatusoutput(commandstring)
    if status != 0:
        raise CommandError
    return output

def parse_version_from_package():
    try:
        pkginfo = os.path.join(glob.glob('*.egg-info')[0],
                                         'PKG-INFO')
    except:
        pkginfo = ''
    
    version_string = ''
    if os.path.exists(pkginfo):
        for line in file(pkginfo):
            if line.find('Version: ') == 0:
                version_string = line.strip().split('Version: ')[1].strip()
        if not version_string:
            version_string = '%s-dev' % version
    else:
        version_string = version
    
    return version_string

def get_version():
    try:
        globalid = execute_command("hg identify -i")
        commitdate = execute_command("hg log -r %s --template '{date|isodatesec}'" % globalid)
        # convert date to UTC unix timestamp, using the date command because python
        # date libraries do not stabilise till about 2.6
        timestamp = int(execute_command('date -d"%s" --utc +%%s' % commitdate))
    
        # finally we have something we can use!
        dt = datetime.datetime.utcfromtimestamp(timestamp)
        datestring = dt.strftime('%Y%m%d%H%M%S')

        version_string = "%s-%s-%s" % (version, datestring, globalid)

    except CommandError, IntegerError:
        version_string = parse_version_from_package()

    return version_string

if __name__ == '__main__':
    print get_version()




gs/content/js/required/browser/javascript/required-min.js
'use strict';jQuery.noConflict();
function GSContentRequiredInterlock(a,b){function c(){var a=0,c=null,l=!0,g=!0;h=[];for(a=0;a<d.length;a++)c=jQuery(d[a]),(l=""!=c.val())||h.push(c),g=g&&l;g?e():(b.attr("disabled","disabled"),f||(f=setInterval(k,m)))}function e(){b.removeAttr("disabled");f&&(clearInterval(f),f=null)}function k(a){c()}function n(a){var b=a=null;a=document.querySelectorAll(".required input");b=document.querySelectorAll(".required input:-webkit-autofill");a.length==b.length&&e()}var d=null,h=null,f=null,m=800;if(null===
b||"undefined"===typeof b)b=a.find('input[type="submit"]');d=a.find(".required input, .required textarea");c();d.keyup(k).on("paste",k);try{document.querySelectorAll("input:-webkit-autofill")&&setTimeout(n,m)}catch(p){}return{required_widgets:function(){return d},unfinished_widgets:function(){return h}}}function GSContentRequired(a,b){var c=null,e=null,c=jQuery(a);null!==b&&"undefined"!==typeof b&&(e=jQuery(b));return GSContentRequiredInterlock(c,e)}
jQuery(window).load(function(){var a=null,b=null,c=null,a=jQuery("form.gs-content-js-required");a.is("*")&&(b=a.attr("data-required-buttons"),"undefined"!==typeof b&&(c=jQuery(b)),GSContentRequiredInterlock(a,c))});


gs/content/js/required/browser/javascript/required.js
'use strict';
// Disable the Submit button of a form if the required widgets are unset.
//
// Copyright © 2013, 2014, 2015 OnlineGroups.net and Contributors.
// All Rights Reserved.
//
// This software is subject to the provisions of the Zope Public License,
// Version 2.1 (ZPL). http://groupserver.org/downloads/license/
//
// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
// WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND
// FITNESS FOR A PARTICULAR PURPOSE.
jQuery.noConflict();

function GSContentRequiredInterlock(form, button) {

    var requiredWidgets = null, unfinished = null, intervalId = null,
        INTERVAL_TIME = 800;

    function check_required_widgets() {
        var i = 0,
            widget = null,
            widgetChecksOk = true,
            checksOk = true;

        unfinished = new Array();  // Clear the unfinished list
        for (i = 0; i < requiredWidgets.length; i++) {
            widget = jQuery(requiredWidgets[i]);
            // The following does not catch radio buttons, nor check
            // boxes. However, it does not make a lick of difference, as
            // both always have a default. (If you radio buttons fail to
            // have a default then you are doing it wrong.)
            widgetChecksOk = (widget.val() != '');
            if (!widgetChecksOk) {  // That is a "not", for those at home
                unfinished.push(widget);
            }
            checksOk = checksOk && widgetChecksOk;
        }

        if (checksOk) {
            enable();
        } else {
            button.attr('disabled', 'disabled');
            if (!intervalId) { // That is a "not", for those at home
                intervalId = setInterval(check, INTERVAL_TIME);
            }
        }
    } // check_required_widgets

    function enable() {
        button.removeAttr('disabled');
        if (intervalId) {
            clearInterval(intervalId);
            intervalId = null;
        }
    } // enable

    function check(event) {
        check_required_widgets();
    } // check

    function i_hate_chrome(event) {
        // Check if the required-widgets that Chrome has autofilled is
        // the same length of the required widgets.
        // http://stackoverflow.com/questions/11708092/detecting-browser-autofill
        var required = null, autofilledRequired = null;
        required = document.querySelectorAll('.required input');
        autofilledRequired = document.querySelectorAll(
            '.required input:-webkit-autofill');
        if (required.length == autofilledRequired.length) {
            enable();
        }
    } // i_hate_chrome

    function init() {
        if ((button === null) || (typeof button === 'undefined')) {
            // If the button has not been passed to us, set the default to
            // be the Submit button.
            button = form.find('input[type="submit"]');
        }

        requiredWidgets = form.find('.required input, .required textarea');
        check_required_widgets();
        requiredWidgets
            .keyup(check)
            .on('paste', check);
        try {
            if (document.querySelectorAll('input:-webkit-autofill')) {
                setTimeout(i_hate_chrome, INTERVAL_TIME);
            }
        } catch (e) {
            // Not Google Chrome
        }
    } // init
    init(); // Note the automatic execution

    return {
        required_widgets: function() {return requiredWidgets;},
        unfinished_widgets: function() {return unfinished;}
    };
} // GSContentRequiredInterlock

function GSContentRequired(formId, buttonId) {
    var form = null, button = null;

    function init() {
        form = jQuery(formId);

        if ((buttonId !== null) && (typeof buttonId !== 'undefined')) {
            button = jQuery(buttonId);
        }
    } // init
    init(); // Note the automatic execution

    return GSContentRequiredInterlock(form, button);
} // GSContentRequired

jQuery(window).load(function() {
    var form = null, buttonId = null, button = null, required = null;

    form = jQuery('form.gs-content-js-required');
    if (form.is('*')) {
        buttonId = form.attr('data-required-buttons');
        if (typeof buttonId !== 'undefined') {
            button = jQuery(buttonId);
        }
        required = GSContentRequiredInterlock(form, button);
    }
});
